web/src/fileview/fileview.js

```diff
@@ -195,6 +195,162 @@ function init(initData) {
+    const stringBefore = textBeforeOffset(node, 0, code);
+
+    const rows = stringBefore.split('\n');
+    // The server expects the rows to be zero-indexed.
+    const row = rows.length - 1;
+    const col = rows[row].length;
+
+    const jumpToDefUrl = "/api/v1/langserver/jumptodef?repo_name=" + initData.repo_info.name + "&file_path=" + initData.file_path + "&row=" + row + "&col=" + col;
+    fetch(jumpToDefUrl, { credentials: "same-origin" })
+      .then(function(response) {
+        if (response.ok) {
+          return response.json();
+        } else {
+          node.classList.add('nonhoverable');
+          return null;
+        }
+      })
+      .then(function(resp) {
+        if (resp) {
+          node.classList.add('hoverable');
+          var asLink = document.createElement('a');
+          asLink.setAttribute('href', resp.url);
+
+          node.parentNode.insertBefore(asLink, node);
+          asLink.appendChild(node);
+
+          if (isInBox(mousePositionX, mousePositionY, node.getBoundingClientRect())) {
+              hoverOverNode(node);
+          }
+        }
+      });
+
+    const hoverUrl = "/api/v1/langserver/hover?repo_name=" + initData.repo_info.name + "&file_path=" + initData.file_path + "&row=" + row + "&col=" + col;    
+    fetch(hoverUrl, { credentials: "same-origin" })
+      .then(function(response) {
+        if (response.ok) {
+          return response.json();
+        } else {
+          node.classList.add('nonhoverable');
+          return null;
+        }
+      })
+      .then(function(resp) {
+        if (resp && resp.contents) {
+          var value = null;
+          if (Array.isArray(resp.contents)) {
+            value = resp.contents.map(formatMarkedStringOrMarkupContent).join('\n');
+          } else {
+            value = formatMarkedStringOrMarkupContent(resp.contents);
+          }
+          if (!value) return null;
+          const backticksStripped = value.replace(/```\w*$/gm, '').trim();
+          $(node).attr('title', backticksStripped);
+          $(node).tooltip();
+        }
+      });
+  }
+
+  function formatMarkedStringOrMarkupContent(markedString) {
+    if (markedString.value) {
+      return '```' + markedString.language + '\n' + markedString.value + '\n```';
+    }
+    return markedString;
+  }
+
+  function isInBox(x, y, rect) {
+    return x >= rect.left && x <= rect.right && y >= rect.top && y <= rect.bottom;
+  }
+
+  // When source code is hovered over, highlight/underline any tokens for which
+  // jump-to-definition will work.
+  function onHover(clientX, clientY) {
+    // The source-code consists of a <code id='source-code' class='code-pane'>
+    // containing lots of <span class="token tokentype">token</span>
+    // hoverable text may be surrounded by <span class="hoverable">
+    // text being hovered over is changed to class="hovering"
+    // non-hoverable text is class="nonhoverable"
+    const pos = document.caretRangeFromPoint(clientX, clientY);
```

Comment:
It looks like [caretRangeFromPoint](https://developer.mozilla.org/en-US/docs/Web/API/Document/caretRangeFromPoint) is nonstandard and doesn't exist in Firefox. Could we wrap this with a wrapper that uses either `caretRangeFromPoint` or [`caretPositionFromPoint`](https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/caretPositionFromPoint) (which is standards-track but experimental and only present in Firefox at present, AFAICT), depending on which one is present?

Reply:
Good catch, done.